portfolio/src/main/webapp/script.js

```diff
@@ -28,103 +28,214 @@ function addRandomGreeting() {
-        if (index == 0){
-            alert("index hit 0");
-            break;
+ * Walks the tree and finds matches,
+ * returns index array of matches.
+ * Will optimize to gather straight from tree
+ * rather than push into new arr to save space
+ */
+function htmlWalkerandMatcher(node, input){
+    let all = [];
+
+    for (node= node.firstChild; node; node=node.nextSibling){
+        if (node.nodeType==3){ //if a text node
+            const regex = new RegExp(input, "i");
+            if (regex.test(node.textContent) == true){
+                all.push(node);
+            }
+        } else {
+            all = all.concat(htmlWalkerandMatcher(node, input));
         }
-        index--;
     }
 
-    var endIndex = index+1; //plus 1 to account for bracket
-
-    while (source.charAt(index) != '<'){
-        index--;
-    }
+    return all;
+}
 
-    var startIndex = index;
+/**
+ * Clears the current highlighted text nodes
+ * Will optimize to actually remove the span rather than the class
+ */
+function clear(){
+    let highlightedList = document.querySelectorAll('.highlight');
 
-    return source.substring(startIndex, endIndex)
+    highlightedList.forEach(function(el) {
+        el.replaceWith(el.innerText);
+    });
+    
 }
 
 /**
-  * Gets HTML id from given word
-  */
-function retrieveId(string, source){
-    var searchForId = string.search("id=\"");
-    if (!searchForId){
+ * Handles clearing old nodes and end of search run,
+ * as well as calls the highlighting and scrolling fns
+ */
+function highlightHandler(matches, input, index){
+    if (index === matches.length && index != 0){
+        clear();
+        setHidden('nextbutton', true);
+        match_index = 0; // resetting the index
+        document.body.scrollIntoView({behavior: "smooth"}); // back to the top
         return;
     }
-    const startIndex = searchForId+4;
-    var retString = string.charAt(startIndex);
 
-    for(var i = startIndex+1; string.charAt(i) != "\""; i++){
-        retString += string.charAt(i);
+    if (index > 0){
+        clear();
     }
+    
+    console.log(index);
+    console.log(matches);
+    console.log(matches[index]);
 
-    return retString;
+    // highlight and scroll to every element
+    highlighter(matches[index], input);
+    
+    console.log(matches[index]);
+
+    let els = document.getElementsByClassName('highlight');
+    els[0].scrollIntoView({behavior: "smooth"}); 
+
+    match_index++; // go onto the next value
 }
 
+
 /**
-  * Runs the search by getting the source HTML, getting the appropriate tag, and scrolling towards that id
-  */
-function runSearch(input){
-    var source = retrieveHtml();
+ * Highlights a given text node
+ */
+function highlighter(node, input){
+    const nodeText = node.textContent;
+
+    let foundIndices = findNeedleInHaystack(input, nodeText);
+
+    console.log(foundIndices);
+    if (foundIndices == [] || foundIndices == null){
+        throw "error with found indices";
+    }
```

Comment:
not sure if you were just using this for debugging purposes, but if no indices were found should this function just not highlight anything and continue?

Reply:
Yes. I put this here for debug because if the code got to this point it means that my walker found a match already in this node, and with foundIndices I'm trying to find where in the node my input it. But if my foundIndices can't find anything it meant there was something wrong with my walker which is why I wanted for the code to stop there

Reply:
Or something wrong with my found Indices, so basically i just wanted to know where in the code I had to go fix